core/migration_external_user_id.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2022 Board of Trustees of the University of Illinois.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"fmt"
+	"log"
+
+	"application/driven/storage"
+
+	"go.mongodb.org/mongo-driver/bson"
+)
+
+// MigrateScoreExternalUserIDs migrates all score records to populate external_user_id with AmgUUID from Core BB
+func (app *Application) MigrateScoreExternalUserIDs(orgID string, appID string, batchSize int) error {
+	log.Println("Starting Score External User ID migration...")
+	log.Printf("Parameters: orgID=%s, appID=%s, batchSize=%d", orgID, appID, batchSize)
+
+	// Validate batch size
+	if batchSize <= 0 {
+		batchSize = 100 // Default to 100 if invalid
+		log.Printf("Invalid batch size, using default: %d", batchSize)
+	}
+
+	// Get all scores that don't have external_user_id populated
+	filter := bson.M{
+		"org_id": orgID,
+		"app_id": appID,
+		"$or": []bson.M{
+			{"external_user_id": bson.M{"$exists": false}},
+			{"external_user_id": ""},
+		},
+	}
+
+	scores, err := app.storage.FindScores(filter, nil, nil)
+	if err != nil {
+		return fmt.Errorf("error fetching scores: %v", err)
+	}
+
+	total := len(scores)
+	log.Printf("Found %d scores to migrate", total)
+
+	if total == 0 {
+		log.Println("No scores to migrate")
+		return nil
+	}
+
+	successCount := 0
+	errorCount := 0
+	skippedCount := 0
+
+	// Process in batches to avoid overwhelming the Core BB API
+	for i := 0; i < total; i += batchSize {
+		end := i + batchSize
+		if end > total {
+			end = total
+		}
+
+		batch := scores[i:end]
+		log.Printf("Processing batch %d-%d of %d", i+1, end, total)
+
+		for _, score := range batch {
+			// Skip if already has external_user_id
+			if score.ExternalUserID != "" {
+				skippedCount++
+				continue
+			}
+
+			// Skip if no external_profile_id (Mastodon ID)
+			if score.ExternalProfileID == "" {
+				log.Printf("Score %s has no external_profile_id, skipping", score.ID)
+				skippedCount++
+				continue
+			}
+
+			// Look up the AmgUUID from Core BB using the Mastodon ID
+			accountCriteria := storage.BuildCoreAccountCriteriaByMastodonID(score.ExternalProfileID)
+
+			coreAccounts, err := app.corebb.RetrieveCoreUserAccountByCriteria(accountCriteria, &appID, &orgID)
```

Comment:
I think we may be able to make a change to these core BB requests to get accounts for multiple mastodon IDs in the same response. We should discuss with Stephen after addressing the other feedback here.